### PR TITLE
Revising the data structure of models

### LINKED ISSUE #5: Revise the model data structure
At several places, the data model doesn't follow principles of good practice. Please consider revision of the model. This includes foremostly the switch from mutable objects to immutable objects.

One example: The `ElectricVehicle` model holds a reference to the target POI. Alongside, it holds a reference to the target POI type (which obviously is an attribute of the POI itself). Additionally, it holds a reference to the destination categorical location, which isn't an attribute of the POI (but could be, afaik).

Revise the data models in order to prevent double and maybe corrupt information.

### LINKED ISSUE #34: FileChanger output problems
When using the current FileChanger implementation, the content of some files do not match the file headline.

### LINKED ISSUE #45: Make `ElectricVehicle` and `ChargingStation` completely immutable
Currently, both have mutable fields (e.g. `var storedEnergy`), which could lead to very bad outcomes and also makes testing more difficult than it needs to be.

### PR DESCRIPTION
**Pull request by staudtMarius**
Resolves #5
Resolves #34
Resolves #45

### CODE REVIEW

File: src/main/scala/edu/ie3/mobsim/model/ElectricVehicle.scala [hunk 1]
```diff
@@ -46,21 +46,18 @@ case class ElectricVehicle(
...
-    private var storedEnergy: ComparableQuantity[Energy],
-    private var destinationPoiType: PoiTypeDictionary.Value,
-    private var destinationCategoricalLocation: CategoricalLocationDictionary.Value,
-    private var destinationPoi: PointOfInterest,
-    private var parkingTimeStart: ZonedDateTime,
-    private var departureTime: ZonedDateTime,
+    private val storedEnergy: ComparableQuantity[Energy],
+    private val destinationPoi: PointOfInterest,
+    private val parkingTimeStart: ZonedDateTime,
+    private val departureTime: ZonedDateTime,
```
Comment (sebastian-peter) [inline src/main/scala/edu/ie3/mobsim/model/ElectricVehicle.scala:52]: I really like this so far. Using `val` over `var` is already a big improvement.
Using some properties of Scala case classes, we can push things even further: The getters below are obsolete if we remove the `private` modifiers here. Using case classes, there is no need for getters anymore, since all fields are immutable now. You can even leave out the `val` modifier, since all fields of a case class are immutable per default.

File: src/main/scala/edu/ie3/mobsim/io/geodata/PointOfInterest.scala [hunk 2]
```diff
@@ -43,6 +46,7 @@ import scala.jdk.CollectionConverters._
 final case class PointOfInterest(
     uuid: UUID,
     id: String,
+    poiType: PoiTypeDictionary.Value,
```
Comment (sebastian-peter) [inline src/main/scala/edu/ie3/mobsim/io/geodata/PointOfInterest.scala:49]: Since `poiType` can be completely derived from `categoricalLocation`, once could also just provide a method here instead of saving it as a field.

**Review by sebastian-peter — CHANGES_REQUESTED** [review]
One very minor point: You can make `ElectricVehicle` a `final case class` now as well :)

Overall, there's a big problem we have now. Everywhere that methods like `setFinalDestinationPoi`, `copyWith` etc. are called, it is still assumed that these methods change the value of fields of the given object instead of creating a new object with the changed values. 
You can find examples of this everywhere. One place would be in `TripSimulation#resumeTrip`, where e.g. `ev.setFinalDestinationPoi(None)` is called without handling the return value - it is just assumed that `ev` absorbs the change and we're done.
So lots of restructuring of calling code is necessary here: The return value needs to be handled and needs to replace the old value wherever it was used, etc.

**Comment by sebastian-peter** [comment]
Just a headsup: In #49, I now removed `ChargingStation.homeChargingStationAssignedToPOI` since it was effectively not used.

File: src/main/scala/edu/ie3/mobsim/model/ElectricVehicle.scala (hunk 1 above)
Comment (staudtMarius) [reply src/main/scala/edu/ie3/mobsim/model/ElectricVehicle.scala]: Dome getters are still needed because this class ist an implementation of the ``EvModel`` found in the Simona API.

**Comment by staudtMarius** [comment]
PR #38 should be merge before this PR

File: src/main/scala/edu/ie3/mobsim/model/ElectricVehicle.scala [hunk 3]
```diff
@@ -78,40 +70,12 @@ case class ElectricVehicle(
 
   def getConsumption: ComparableQuantity[SpecificEnergy] = consumption
```
Comment (sebastian-peter) [inline src/main/scala/edu/ie3/mobsim/model/ElectricVehicle.scala:71]: This getter has also become unnecessary, one can use `ev.consumption` instead

File: src/main/scala/edu/ie3/mobsim/MobilitySimulator.scala [hunk 4]
```diff
@@ -574,6 +582,20 @@ final class MobilitySimulator(
...
+    val movementMap: Map[UUID, ElectricVehicle] = movements.map { movement =>
+      movement.ev.uuid -> movement.ev
+    }.toMap
+
+    electricVehicles = electricVehicles.map { ev =>
+      if (movementMap.contains(ev.uuid)) {
+        movementMap(ev.uuid)
+      } else {
+        ev
+      }
```
Comment (sebastian-peter) [inline src/main/scala/edu/ie3/mobsim/MobilitySimulator.scala:596]: This can be shortened:
```suggestion
      movementMap.getOrElse(ev.uuid, ev)
```

File: src/main/scala/edu/ie3/mobsim/MobilitySimulator.scala [hunk 5]
```diff
@@ -574,6 +582,20 @@ final class MobilitySimulator(
 
     timeUntilNextEvent
   }
+
+  private def updateElectricVehicles(movements: Seq[Movement]): Unit = {
+    val movementMap: Map[UUID, ElectricVehicle] = movements.map { movement =>
```
Comment (sebastian-peter) [inline src/main/scala/edu/ie3/mobsim/MobilitySimulator.scala:587]: I like building this map here, very efficient :)
Just a minor point: You don't need to assign types (`Map[UUID, ElectricVehicle]` here) to local vals.

File: src/test/scala/edu/ie3/mobsim/model/ElectricVehicleSpec.scala [hunk 6]
```diff
@@ -44,38 +49,36 @@ class ElectricVehicleSpec extends UnitSpec with ElectricVehicleTestData {
...
                 chargingAtHomePossible,
                 chosenChargingStation,
                 chargingAtSimona,
-                finalDestinationPoiType,
                 finalDestinationPoi,
                 remainingDistanceAfterChargingHub,
                 chargingPricesMemory
               ) =>
             simulationStart shouldBe givenSimulationStart
+            uuid shouldBe ev.getUuid
```
Comment (sebastian-peter) [inline src/test/scala/edu/ie3/mobsim/model/ElectricVehicleSpec.scala:63]: This test does not make sense, as it's always true, per definition.

File: src/test/scala/edu/ie3/mobsim/model/ElectricVehicleSpec.scala [hunk 7]
```diff
@@ -44,38 +49,36 @@ class ElectricVehicleSpec extends UnitSpec with ElectricVehicleTestData {
...
-            workPoi shouldBe givenWorkPoi
+            homePoi shouldBe ev.homePoi
+            workPoi shouldBe ev.workPoi
             storedEnergy shouldBe givenModel.capacity
             chargingAtSimona shouldBe false
-            destinationPoiType shouldBe PoiTypeDictionary.HOME
-            destinationCategoricalLocation shouldBe CategoricalLocationDictionary.HOME
+            ev.destinationPoi.getPoiType shouldBe PoiTypeDictionary.HOME
+            ev.destinationPoi.categoricalLocation shouldBe CategoricalLocationDictionary.HOME
             destinationPoi shouldBe givenHomePoi
```
Comment (sebastian-peter) [inline src/test/scala/edu/ie3/mobsim/model/ElectricVehicleSpec.scala:76]: These two lines are obsolete, as `destinationPoi` is already checked in the next line

File: src/test/scala/edu/ie3/mobsim/model/ElectricVehicleSpec.scala [hunk 8]
```diff
@@ -44,38 +49,36 @@ class ElectricVehicleSpec extends UnitSpec with ElectricVehicleTestData {
...
             model shouldBe "cool_producer cool_model"
             batteryCapacity shouldBe givenModel.capacity
             acChargingPower shouldBe givenModel.acPower
             dcChargingPower shouldBe givenModel.dcPower
             consumption shouldBe givenModel.consumption
-            homePoi shouldBe givenHomePoi
-            workPoi shouldBe givenWorkPoi
+            homePoi shouldBe ev.homePoi
+            workPoi shouldBe ev.workPoi
             storedEnergy shouldBe givenModel.capacity
```
Comment (sebastian-peter) [inline src/test/scala/edu/ie3/mobsim/model/ElectricVehicleSpec.scala:72]: These do not make sense either. Why did you change these lines?

File: src/test/scala/edu/ie3/mobsim/model/ElectricVehicleSpec.scala [hunk 9]
```diff
@@ -272,10 +275,93 @@ class ElectricVehicleSpec extends UnitSpec with ElectricVehicleTestData {
...
             evs should have size targetAmount
             evs.count(
-              _.isChargingAtHomePossible
+              _.chargingAtHomePossible
             ) shouldBe expectedAmountOfHomeCharging
         }
       }
     }
+
+    "An electricVehicle" should {
```
Comment (sebastian-peter) [inline src/test/scala/edu/ie3/mobsim/model/ElectricVehicleSpec.scala:284]: I like that you're testing the copy methods here. It might seem trivial at first, but I think we can get some value here in the long term

File: src/test/scala/edu/ie3/mobsim/model/ElectricVehicleSpec.scala [hunk 10]
```diff
@@ -13,28 +13,33 @@ import edu.ie3.mobsim.io.geodata.PoiEnums.{
...
           givenHomePoi,
           givenWorkPoi,
           givenSimulationStart,
           givenFirstDeparture,
           isChargingAtHomePossible = true
-        ) match {
+        )
+
+        ev match {
           case ElectricVehicle(
```
Comment (sebastian-peter) [inline src/test/scala/edu/ie3/mobsim/model/ElectricVehicleSpec.scala:40]: As described below, I don't think you need the assignment to variable `ev`

File: src/test/scala/edu/ie3/mobsim/model/ChargingBehaviorTestData.scala [hunk 11]
```diff
@@ -18,45 +14,34 @@ trait ChargingBehaviorTestData extends TripSimulationTestData {
 
   protected def evLowSoC: ElectricVehicle = ev1.copyWith(
```
Comment (sebastian-peter) [inline src/test/scala/edu/ie3/mobsim/model/ChargingBehaviorTestData.scala:15]: You can turn all those `def`s back into `val`s now, should speed up tests a tiny bit :) Same for `TripSimulationTestData`

File: src/test/scala/edu/ie3/mobsim/io/geodata/PoiEnumsSpec.scala [hunk 12]
```diff
@@ -10,9 +10,13 @@ import edu.ie3.mobsim.io.geodata.PoiEnums.{
...
   PoiTypeDictionary
 }
+import edu.ie3.mobsim.model.{ElectricVehicleTestData, TripSimulationTestData}
 import edu.ie3.test.common.UnitSpec
 
-class PoiEnumsSpec extends UnitSpec {
+class PoiEnumsSpec
+    extends UnitSpec
+    with ElectricVehicleTestData
+    with TripSimulationTestData {
```
Comment (sebastian-peter) [inline src/test/scala/edu/ie3/mobsim/io/geodata/PoiEnumsSpec.scala:20]: What do you need these traits for here?

File: src/test/scala/edu/ie3/mobsim/model/ElectricVehicleTestData.scala [hunk 13]
```diff
@@ -87,4 +87,23 @@ trait ElectricVehicleTestData {
     1,
     homeChargingStationAssignedToPOI = true
   )
+
+  val evWithHomeCharging: ElectricVehicle = ElectricVehicle.buildEv(
```
Comment (sebastian-peter) [inline src/test/scala/edu/ie3/mobsim/model/ElectricVehicleTestData.scala:91]: Also make 'em protected?

File: src/test/scala/edu/ie3/mobsim/model/TripSimulationSpec.scala [hunk 14]
```diff
@@ -20,7 +22,7 @@ import java.time.ZonedDateTime
 import javax.measure.quantity.Energy
 import scala.collection.mutable
 
-class TripSimulationSpec extends UnitSpec with ChargingBehaviorTestData {
+class TripSimulationSpec extends UnitSpec with IoUtilsTestData {
```
Comment (sebastian-peter) [inline src/test/scala/edu/ie3/mobsim/model/TripSimulationSpec.scala:25]: Similar notes here compared to `ElectricVehicleSpec`:
- Assignments such as `val ev: ElectricVehicle = evAtChargingHub` are unnecessary
- Comparing `ev.getUuid` to itself does not make sense
- `destinationPoi` only needs to be checked once

File: src/test/scala/edu/ie3/mobsim/model/TripSimulationSpec.scala [hunk 15]
```diff
@@ -80,14 +79,14 @@ class TripSimulationSpec extends UnitSpec with ChargingBehaviorTestData {
...
-          destinationCategoricalLocation shouldBe CategoricalLocationDictionary.CHARGING_HUB_TOWN
+          destinationPoi.getPoiType shouldBe PoiTypeDictionary.CHARGING_HUB_TOWN
+          destinationPoi.categoricalLocation shouldBe CategoricalLocationDictionary.CHARGING_HUB_TOWN
           destinationPoi shouldBe charging_hub_townPoi
           parkingTimeStart shouldBe simulationStart.plusMinutes(1)
           departureTime shouldBe simulationStart.plusHours(4).plusMinutes(33)
           chargingAtHomePossible shouldBe true
           chosenChargingStation shouldBe None
-          finalDestinationPoiType shouldBe None
+          finalDestinationPoi.map(_.getPoiType) shouldBe None
```
Comment (sebastian-peter) [inline src/test/scala/edu/ie3/mobsim/model/TripSimulationSpec.scala:89]: Unnecessary mapping here, you could leave this as it was before

File: src/main/scala/edu/ie3/mobsim/model/TripSimulation.scala [hunk 16]
```diff
@@ -1017,17 +974,14 @@ object TripSimulation extends LazyLogging {
...
       plannedDestinationPoi: PointOfInterest,
       plannedParkingTimeStart: ZonedDateTime,
       plannedDepartureTime: ZonedDateTime
   ): ElectricVehicle = {
 
     /* Because there is no stop at a charging hub, no trip values need to be saved */
-    ev.setFinalDestinationPoiType(None)
-    ev.setFinalDestinationPoi(None)
-    ev.setRemainingDistanceAfterChargingHub(None)
+    var updatedEv: ElectricVehicle = ev.setFinalDestinationPoi(None)
```
Comment (sebastian-peter) [inline src/main/scala/edu/ie3/mobsim/model/TripSimulation.scala:983]: You can chain these like `ev.setFinalDestinationPoi(None).setRemainingDistanceAfterChargingHub(None)`. This way, you don't need vars

File: src/main/scala/edu/ie3/mobsim/model/TripSimulation.scala [hunk 17]
```diff
@@ -1042,14 +996,13 @@ object TripSimulation extends LazyLogging {
     }
 
     /* Create updated EV */
-    ev.copyWith(
+    updatedEv = updatedEv.copyWith(
```
Comment (sebastian-peter) [inline src/main/scala/edu/ie3/mobsim/model/TripSimulation.scala:999]: Also unnecessary, you can just return what `updatedEv.copyWith` returns

File: src/main/scala/edu/ie3/mobsim/model/TripSimulation.scala [hunk 18]
```diff
@@ -375,20 +364,15 @@ object TripSimulation extends LazyLogging {
...
     * @param distance
     *   Distance to be driven
     */
   private final case class TargetProperties(
-      poiType: PoiTypeDictionary.Value,
-      categoricalLocation: CategoricalLocationDictionary.Value,
       poi: PointOfInterest,
-      distance: ComparableQuantity[Length]
+      distance: ComparableQuantity[Length],
+      updatedEv: ElectricVehicle
```
Comment (sebastian-peter) [inline src/main/scala/edu/ie3/mobsim/model/TripSimulation.scala:375]: Add ScalaDoc for this new field?

File: src/main/scala/edu/ie3/mobsim/model/TripSimulation.scala [hunk 19]
```diff
@@ -903,21 +866,19 @@ object TripSimulation extends LazyLogging {
       newParkingTimeStart.plusMinutes(newParkingTime)
 
     /* Save parameters of the originally planned trip for the next trip simulation after charging */
-    ev.setRemainingDistanceAfterChargingHub(
+    var updatedEv: ElectricVehicle = ev.setRemainingDistanceAfterChargingHub(
```
Comment (sebastian-peter) [inline src/main/scala/edu/ie3/mobsim/model/TripSimulation.scala:869]: These could also be chained

File: src/main/scala/edu/ie3/mobsim/model/ChargingStation.scala [hunk 20]
```diff
@@ -24,7 +24,7 @@ case class ChargingStation(
     private val evcsType: ChargingPointType,
     private val evcsLocationType: EvcsLocationType,
     private val chargingPoints: Int,
-    private var homeChargingStationAssignedToPOI: Boolean
+    private val homeChargingStationAssignedToPOI: Boolean
 ) {
```
Comment (sebastian-peter) [inline src/main/scala/edu/ie3/mobsim/model/ChargingStation.scala:28]: You could make all fields of `ChargingStation` public by removing the `private val` properties. Then, you could also remove the methods below

**Review by sebastian-peter — CHANGES_REQUESTED** [review]
Very good job making  these data classes immutable! I've got some minor comments for now.

File: src/test/scala/edu/ie3/mobsim/io/geodata/PoiEnumsSpec.scala (hunk 12 above)
Comment (staudtMarius) [reply src/test/scala/edu/ie3/mobsim/io/geodata/PoiEnumsSpec.scala:20]: There are unnecessary. I will remove them.

File: src/test/scala/edu/ie3/mobsim/model/TripSimulationSpec.scala (hunk 15 above)
Comment (staudtMarius) [reply src/test/scala/edu/ie3/mobsim/model/TripSimulationSpec.scala:89]: The old way does not work anymore, because the field finalDestinationPoiType was removed from ``ElectricVehicle``.

File: src/main/scala/edu/ie3/mobsim/model/ChargingBehavior.scala [hunk 21]
```diff
@@ -74,39 +75,49 @@ object ChargingBehavior extends LazyLogging {
...
-          createRatingsForChargingStations(
+          val ratingMap = createRatingsForChargingStations(
             ev,
             currentPricesAtChargingStations,
             currentlyAvailableChargingPoints,
             maxDistance
-          ).maxByOption(_._2).map(_._1)
+          )
+          (
+            ratingMap.maxByOption(_._2).map(_._1),
```
Comment (sebastian-peter) [inline src/main/scala/edu/ie3/mobsim/model/ChargingBehavior.scala:105]: Please don't use the shorthand tuple methods `_1` and `_2`, but rather go for a case-mapping like this:

```suggestion
            ratingMap
              .maxByOption { case (_, rating) => rating }
              .map { case (cs, _) => cs },
```

Yes, this way we have longer code, but it's actually easier to grasp on first sight what is happening.

File: src/main/scala/edu/ie3/mobsim/model/ChargingBehavior.scala [hunk 22]
```diff
@@ -74,39 +75,49 @@ object ChargingBehavior extends LazyLogging {
...
-          .map(_.getUuid)
+        (
+          ev.destinationPoi.nearestChargingStations.keys.headOption
+            .map(_.uuid),
+          None
+        )
       } else {
         /* Update charging prices memory of EV to have a reference for the prices of specific charging stations */
-        ev.getDestinationPoi.nearestChargingStations.keys.foreach { cs =>
+        val priceQueue: mutable.Queue[Double] = mutable.Queue.empty
```
Comment (sebastian-peter) [inline src/main/scala/edu/ie3/mobsim/model/ChargingBehavior.scala:85]: You can easily solve this in a functional style without using mutable objects :)
Hint: for that you'd probably need to call `map` instead of `foreach` below.

File: src/main/scala/edu/ie3/mobsim/model/ElectricVehicle.scala [hunk 23]
```diff
@@ -131,51 +90,58 @@ case class ElectricVehicle(
...
   def setRemainingDistanceAfterChargingHub(
       remainingDistance: Option[ComparableQuantity[Length]]
-  ): Unit = {
-    remainingDistanceAfterChargingHub = remainingDistance
+  ): ElectricVehicle = {
+    copy(remainingDistanceAfterChargingHub = remainingDistance)
   }
 
-  def updateChargingPricesMemory(newPrice: Double): mutable.Queue[Double] = {
+  def updateChargingPricesMemory(
```
Comment (sebastian-peter) [inline src/main/scala/edu/ie3/mobsim/model/ElectricVehicle.scala:130]: I like that you turned the `chargingPricesMemory` into an immutable queue. Please also only accept immutable queues as `newPrices` here. Once all data structures here have been made immutable, you can also just use `Queue[Double]` without the prefix.

File: src/main/scala/edu/ie3/mobsim/model/ChargingBehavior.scala [hunk 24]
```diff
@@ -120,32 +131,26 @@ object ChargingBehavior extends LazyLogging {
...
+        (
+          ev.chargingAtHomePossible,
+          categoricalLocation == CategoricalLocationDictionary.HOME
+            || categoricalLocation == CategoricalLocationDictionary.WORK
+        ) match {
+          case (true, true)   => (0.4, 0.85)
+          case (true, false)  => (0.3, 0.5)
+          case (false, true)  => (0.4, 0.85)
+          case (false, false) => (0.3, 0.75)
         }
```
Comment (sebastian-peter) [inline src/main/scala/edu/ie3/mobsim/model/ChargingBehavior.scala:154]: I like this a lot better than what we had before here!

File: src/main/scala/edu/ie3/mobsim/model/TripSimulation.scala [hunk 25]
```diff
@@ -584,11 +546,11 @@ object TripSimulation extends LazyLogging {
       ],
       depth: Int = 0,
       maxDepth: Int = 4
-  ): (PoiEnums.CategoricalLocationDictionary.Value, PointOfInterest) = {
+  ): PointOfInterest = {
```
Comment (sebastian-peter) [inline src/main/scala/edu/ie3/mobsim/model/TripSimulation.scala:549]: Please also adapt the ScalaDoc of this method to the changes you made here

File: src/main/scala/edu/ie3/mobsim/model/TripSimulation.scala [hunk 26]
```diff
@@ -533,14 +495,14 @@ object TripSimulation extends LazyLogging {
           PointOfInterest
         ]
       ]
-  ): (PoiEnums.CategoricalLocationDictionary.Value, PointOfInterest) = {
+  ): PointOfInterest = {
```
Comment (sebastian-peter) [inline src/main/scala/edu/ie3/mobsim/model/TripSimulation.scala:498]: ScalaDoc

File: src/main/scala/edu/ie3/mobsim/MobilitySimulator.scala [hunk 27]
```diff
@@ -302,9 +307,12 @@ final class MobilitySimulator(
...
         case result @ Some(_) =>
-          ev.setChargingAtSimona(false)
-          ev.setChosenChargingStation(None)
-          result
+          val updatedEv: ElectricVehicle =
+            ev.removeChargingAtSimona().setChosenChargingStation(None)
+
+          val cs: UUID = result.value._1
+
+          Some(cs, MobilitySimulator.Movement(cs, updatedEv))
```
Comment (sebastian-peter) [inline src/main/scala/edu/ie3/mobsim/MobilitySimulator.scala:315]: This method needs some more refactoring imho. The movement created in line 303 just goes to waste and can be removed. In turn, all of these ev updates can go into the mapping part from line 302-306. Hit me up if this needs discussion :)

File: src/main/scala/edu/ie3/mobsim/MobilitySimulator.scala [hunk 28]
```diff
@@ -411,38 +419,45 @@ final class MobilitySimulator(
...
+            .setChargingAtSimona()
+            .setChosenChargingStation(Some(cs))
 
-        Some((cs, Movement(cs, ev)))
-      } else {
+          logger.debug(
+            s"${ev.getId} starts charging at $cs."
+          )
+
+          Some((cs, Movement(cs, updatedEv)))
```
Comment (sebastian-peter) [inline src/main/scala/edu/ie3/mobsim/MobilitySimulator.scala:445]: Since the `cs` UUID is already saved within movement, this method can easily return `Future[Option[Movement]]` only

File: src/test/scala/edu/ie3/mobsim/model/ElectricVehicleTestData.scala [hunk 29]
```diff
@@ -87,4 +87,24 @@ trait ElectricVehicleTestData {
...
+  protected val evWithoutHomeCharging: ElectricVehicle =
+    ElectricVehicle.buildEv(
+      "test_car",
+      givenModel,
+      givenHomePoi,
+      givenWorkPoi,
+      givenSimulationStart,
+      givenFirstDeparture,
+      isChargingAtHomePossible = false
+    )
```
Comment (sebastian-peter) [inline src/test/scala/edu/ie3/mobsim/model/ElectricVehicleTestData.scala:109]:
```suggestion
evWithHomeCharging.copy(chargingAtHomePossible = false)
```

File: src/test/scala/edu/ie3/mobsim/utils/IoUtilsSpec.scala [hunk 30]
```diff
@@ -0,0 +1,205 @@
...
+        } else {
+          (
+            firstEv.destinationPoi.categoricalLocation.toString,
+            firstEv.destinationPoi.toString
+          )
+        }
+
+      val compareString: String = (s"$uuid," +
+        s"${firstEv.uuid.toString}," +
+        s"$location,").replaceAll(",", ";") +
```
Comment (sebastian-peter) [inline src/test/scala/edu/ie3/mobsim/utils/IoUtilsSpec.scala:195]: Why are you placing a `,` in the String and the replace it with `;`? Same also above

File: src/test/scala/edu/ie3/mobsim/utils/IoUtilsTestData.scala [hunk 31]
```diff
@@ -0,0 +1,48 @@
...
+  protected val secondEv: ElectricVehicle = ev2
+
+  protected val evSet: Set[ElectricVehicle] = Seq(firstEv, secondEv).toSet
+
+  protected val poiMap
+      : Map[CategoricalLocationDictionary.Value, Set[PointOfInterest]] = {
+    val poiSet: Set[PointOfInterest] = Seq(charging_hub_townPoi).toSet
+
+    poiSet.groupBy(_.categoricalLocation).map { case (catLoc, poi) =>
+      catLoc -> TreeSet.from(poi)
```
Comment (sebastian-peter) [inline src/test/scala/edu/ie3/mobsim/utils/IoUtilsTestData.scala:45]: What is the TreeSet good for?

**Review by sebastian-peter — CHANGES_REQUESTED** [review]

File: src/test/scala/edu/ie3/mobsim/utils/IoUtilsTestData.scala (hunk 31 above)
Comment (staudtMarius) [reply src/test/scala/edu/ie3/mobsim/utils/IoUtilsTestData.scala:45]: Some information are seperated by ``,`` but need to be seperated by ``;``. Therefor I am using the replaceAll method.

File: src/test/scala/edu/ie3/mobsim/utils/IoUtilsSpec.scala (hunk 30 above)
Comment (staudtMarius) [reply src/test/scala/edu/ie3/mobsim/utils/IoUtilsSpec.scala:195]: Some information are seperated by ``,`` but need to be seperated by ``;``. Therefor I am using the replaceAll method.

File: src/main/scala/edu/ie3/mobsim/model/ChargingBehavior.scala [hunk 32]
```diff
@@ -74,39 +75,51 @@ object ChargingBehavior extends LazyLogging {
...
-          currentPricesAtChargingStations
-            .get(cs.getUuid)
-            .map(ev.updateChargingPricesMemory)
-        }
+
+        val prices = immutable.Queue.from(
+          ev.destinationPoi.nearestChargingStations.keys.map { cs =>
+            currentPricesAtChargingStations(cs.uuid)
+          }
+        )
```
Comment (sebastian-peter) [inline src/main/scala/edu/ie3/mobsim/model/ChargingBehavior.scala:90]: Could be simplified with the `to` method:

```suggestion
        val prices = ev.destinationPoi.nearestChargingStations.keys
          .map { cs =>
            currentPricesAtChargingStations(cs.uuid)
          }
          .to(immutable.Queue)
```

File: src/main/scala/edu/ie3/mobsim/MobilitySimulator.scala [hunk 33]
```diff
@@ -290,26 +298,25 @@ final class MobilitySimulator(
...
+          /* Register departure */
+          logger.debug(
+            s"${ev.getId} departs from $cs."
+          )
+
+          val updatedEv: ElectricVehicle =
+            ev.removeChargingAtSimona().setChosenChargingStation(None)
+
+          Some(MobilitySimulator.Movement(cs, updatedEv))
+        case None =>
```
Comment (sebastian-peter) [inline src/main/scala/edu/ie3/mobsim/MobilitySimulator.scala:315]: Yes! That's what I meant.

File: src/main/scala/edu/ie3/mobsim/model/ElectricVehicle.scala [hunk 34]
```diff
@@ -131,51 +90,55 @@ case class ElectricVehicle(
...
+
     /* keep maximum of 20 last known prices */
-    while (chargingPricesMemory.size >= 20) {
-      chargingPricesMemory.dequeue()
+    while (queue.size >= 20) {
+      queue = queue.dequeue._2
     }
-    chargingPricesMemory += newPrice
+
+    copy(chargingPricesMemory = Queue.empty.enqueueAll(queue))
```
Comment (sebastian-peter) [inline src/main/scala/edu/ie3/mobsim/model/ElectricVehicle.scala:141]: This can be solved in a simpler way and without `var`s like this:

```suggestion
    val newPricesQueue = (chargingPricesMemory ++ newPrices).takeRight(20)

    copy(chargingPricesMemory = newPricesQueue)
```

File: src/main/scala/edu/ie3/mobsim/model/ElectricVehicle.scala [hunk 35]
```diff
@@ -473,10 +432,9 @@ case object ElectricVehicle extends LazyLogging {
       chargingAtHomePossible = isChargingAtHomePossible,
       chosenChargingStation = None,
       chargingAtSimona = false,
-      finalDestinationPoiType = None,
       finalDestinationPoi = None,
       remainingDistanceAfterChargingHub = None,
-      chargingPricesMemory = mutable.Queue[Double]()
+      chargingPricesMemory = immutable.Queue[Double]()
```
Comment (sebastian-peter) [inline src/main/scala/edu/ie3/mobsim/model/ElectricVehicle.scala:437]: Nicer like this (since you already imported the immutable Queue anyways):

```suggestion
      chargingPricesMemory = Queue.empty[Double]
```

File: src/main/scala/edu/ie3/mobsim/model/ElectricVehicle.scala [hunk 36]
```diff
@@ -9,108 +9,69 @@ package edu.ie3.mobsim.model
...
 import java.time.ZonedDateTime
 import java.util.UUID
-import edu.ie3.simona.api.data.ev.model.EvModel
-import edu.ie3.util.quantities.PowerSystemUnits
-import tech.units.indriya.quantity.Quantities
-
 import javax.measure.quantity.{Energy, Length, Power}
-import scala.collection.immutable.SortedSet
-import scala.collection.mutable
+import scala.collection.immutable
```
Comment (sebastian-peter) [inline src/main/scala/edu/ie3/mobsim/model/ElectricVehicle.scala:29]: Unnecessary import (after all comments in this class are solved)

File: src/test/scala/edu/ie3/mobsim/utils/IoUtilsSpec.scala [hunk 37]
```diff
@@ -0,0 +1,205 @@
...
+        } else {
+          (
+            firstEv.destinationPoi.categoricalLocation.toString,
+            firstEv.destinationPoi.toString
+          )
+        }
+
+      val compareString: String = (s"$uuid," +
+        s"${firstEv.uuid.toString};" +
+        s"$location;").replaceAll(",", ";") +
```
Comment (sebastian-peter) [inline src/test/scala/edu/ie3/mobsim/utils/IoUtilsSpec.scala:195]: Ok good, but now the `replaceAll` call is not necessary anymore?

File: src/test/scala/edu/ie3/mobsim/model/ElectricVehicleSpec.scala [hunk 38]
```diff
@@ -272,10 +265,92 @@ class ElectricVehicleSpec extends UnitSpec with ElectricVehicleTestData {
...
+          evNoQueue.updateChargingPricesMemory(queue)
+        updatedEv.chargingPricesMemory shouldBe queue
+
+        for (n <- 11 to 25) {
+          queue.enqueue(n.doubleValue())
+        }
+
+        updatedEv = updatedEv.updateChargingPricesMemory(queue)
+        updatedEv.chargingPricesMemory shouldBe queue.drop(7)
+      }
```
Comment (sebastian-peter) [inline src/test/scala/edu/ie3/mobsim/model/ElectricVehicleSpec.scala:353]: Since the queues are now immutable, test now does not work anymore.

Please use `(1 to 20).to(immutable.Queue)` or the like to set up the test queues.

**Review by sebastian-peter — CHANGES_REQUESTED** [review]
I think we're getting really close to the merge. Not quite there yet though.

File: src/main/scala/edu/ie3/mobsim/model/ChargingBehavior.scala [hunk 39]
```diff
@@ -74,39 +75,51 @@ object ChargingBehavior extends LazyLogging {
...
-            .map(ev.updateChargingPricesMemory)
-        }
+
+        val prices = ev.destinationPoi.nearestChargingStations.keys
+          .map { cs =>
+            currentPricesAtChargingStations(cs.uuid)
+          }
+          .to(Queue)
+
+        val evWithUpdatedPriceMemory = ev.updateChargingPricesMemory(prices)
```
Comment (sebastian-peter) [inline src/main/scala/edu/ie3/mobsim/model/ChargingBehavior.scala:92]: `evWithUpdatedPriceMemory` is not used with `doesEvWantToCharge` and `createRatingsForChargingStations`, which I believe can lead to wrong results.

Please fix this and also create a test case that fails now but succeeds once this is fixed, as this is very important to function the same way as before.

**Review by sebastian-peter — CHANGES_REQUESTED** [review]
I just made some minor updates to some lines relating to code style and tests, so that we can get this merged.

Now I found something that could become a major bug, related to the changes of this PR. In general, please make sure that tests are covering a wide enough range of cases so that these problems surface immediately.

**Review by sebastian-peter — APPROVED** [review]
Thanks a lot, and congrats for getting this done!
